### PR TITLE
Adds persistence of API results to make app smoother

### LINKED ISSUE #33: Cache previous state in case of no internet connection
Things like the user's QR code will never change so it would be nice to persist it in case there's a problem with the user's internet connection.

### PR DESCRIPTION
**Pull request by patrickfeltes**
- Closes #33 

- This PR implements the Repository pattern with the android Room database. Basically, the repository pattern abstracts away where the data comes from. When fetching data, the repository returns a LiveData object pointing to the local database and then calls the API in the background, if necessary (right now data goes stale after 2 minutes). Once the API call finishes, the data in the database is updated, triggering the LiveData. This will automatically update the UI since we have observers in the Fragments/Activities.
- Added try/catch to each retrofit API call so that it won't crash when we don't have internet. Instead it just shows whatever is in the local database.
- Reference for Repository/Room if you're interested: [https://developer.android.com/jetpack/docs/guide](https://developer.android.com/jetpack/docs/guide)

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -28,13 +29,20 @@ android {
 }
 
 dependencies {
+    def room_version = "1.1.1"
```
Comment (bstrauch24) [inline app/build.gradle]: You should put this variable right before line 43 where it's being used

File: app/src/main/java/org/hackillinois/android/App.java [hunk 2]
```diff
@@ -13,9 +16,11 @@
 
 public class App extends Application {
     private static API api;
+    private static Database database;
 
     public void onCreate() {
         super.onCreate();
+        database = Room.databaseBuilder(getApplicationContext(), Database.class, "local-db").fallbackToDestructiveMigration().build();
```
Comment (bstrauch24) [inline app/src/main/java/org/hackillinois/android/App.java]: When using the builder pattern, you should put each additional method call on a new line

File: app/src/main/java/org/hackillinois/android/database/dao/AttendeeDao.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import android.arch.lifecycle.LiveData
+import android.arch.persistence.room.*
+import org.hackillinois.android.database.entity.Attendee
+
+@Dao
+interface AttendeeDao {
+    @Query("SELECT * FROM attendees LIMIT 1")
+    fun getAttendee(): LiveData<Attendee>
+
+    @Query("SELECT * FROM attendees WHERE last_refreshed > :lastRefreshMax LIMIT 1")
```
Comment (bstrauch24) [inline app/src/main/java/org/hackillinois/android/database/dao/AttendeeDao.kt]: This SQL stuff is pretty cool

**Comment by patrickfeltes** [comment]
I didn't look at any of the caching code from last year, so I didn't even know that OkHttp had a caching mechanism.

I think the main benefit from using room is that with it we can make our LiveData objects references to the database memory, and then all of the updating of those is handled when we write to the database. Mostly, I was just going off the android architecture guide, but it is probably worth considering the OkHttp cache if it's simpler.

**Comment by patrickfeltes** [comment]
Another nice thing about using Room for caching is that when we start up the app, we automatically populate the views with our local data storage, and then it is updated once the new network request is finished, which gets rid of the delay in populating the views.

I'm not sure if the method they used last year will do that, since it appears to make a request and only use the local storage if there is no internet. Really depends on what we want from a user experience though, so let me know what you think.

**Pull request merged**